### PR TITLE
fix asan fail for `cast_decimal_as_json` and `cast_time_as_json`

### LINKED ISSUE #8427: asan fail for cast_decimal_as_json and cast_time_as_json
https://ci.pingcap.net/blue/organizations/jenkins/tiflash-sanitizer-daily/detail/tiflash-sanitizer-daily/1544/pipeline/
```
[2023-11-27T19:43:15.895Z] FAILED TESTS (29/2386):
[2023-11-27T19:43:15.895Z]     2525 ms: /tiflash/gtests_dbms TestCastAsJson.CastDecimalAsJson
[2023-11-27T19:43:15.895Z]     2722 ms: /tiflash/gtests_dbms TestCastAsJson.CastTimeAsJson
```

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #8427

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by SeaRise** [comment]
> /run-sanitizer-test asan

https://ci.pingcap.net/blue/organizations/jenkins/tiflash-sanitizer-daily/detail/tiflash-sanitizer-daily/1548/pipeline

**Comment by SeaRise** [comment]
> > /run-sanitizer-test asan
> 
> https://ci.pingcap.net/blue/organizations/jenkins/tiflash-sanitizer-daily/detail/tiflash-sanitizer-daily/1548/pipeline

fixed

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Functions/FunctionsJson.h [hunk 1]
```diff
@@ -787,11 +787,12 @@ class FunctionCastDecimalAsJson : public IFunction
...
         for (size_t i = 0; i < column_from->size(); ++i)
         {
-            const auto & field = (*column_from)[i].template safeGet<DecimalField<FromType>>();
             // same as https://github.com/pingcap/tidb/blob/90628349860718bb84c94fe7dc1e1f9bd9da4348/pkg/expression/builtin_cast.go#L854-L865
             // https://github.com/pingcap/tidb/issues/48796
             // TODO `select json_type(cast(1111.11 as json))` should return `DECIMAL`, we return `DOUBLE` now.
-            JsonBinary::appendNumber(write_buffer, static_cast<Float64>(field));
+            JsonBinary::appendNumber(
+                write_buffer,
+                static_cast<Float64>((*column_from)[i].template safeGet<DecimalField<FromType>>()));
```
Comment (SeaRise) [inline dbms/src/Functions/FunctionsJson.h:795]: I don’t know why, but it works....

File: dbms/src/Functions/tests/gtest_cast_as_json.cpp [hunk 2]
```diff
@@ -322,20 +322,20 @@ try
...
     // DataTypeMyDateTime
+    // Only raw function test is tested, so input_tidb_tp is always nullptr and only the case of TiDB::TypeTimestamp is tested here.
     {
         auto data_col_ptr
             = createColumn<DataTypeMyDateTime::FieldType>(
                   {MyDateTime(2023, 1, 2, 3, 4, 5, 6).toPackedUInt(), MyDateTime(0, 0, 0, 0, 0, 0, 0).toPackedUInt()})
                   .column;
         ColumnWithTypeAndName input(data_col_ptr, datetime_type_ptr, "");
-        auto res = executeFunctionWithCast("cast_time_as_json", {input});
+        auto res = executeFunctionWithCast<true>("cast_time_as_json", {input});
```
Comment (SeaRise) [inline dbms/src/Functions/tests/gtest_cast_as_json.cpp:332]: Use `raw_funtion_test = true` to skip constructing `tipb::FieldType`

**Pull request merged**